### PR TITLE
Migrate away from `@testing-library/react-hooks`

### LINKED ISSUE #10333: Migrate tests from `@testing-library/react-hooks` to `@testing-library/react`
React hooks testing library does not have React 18 support and has [migrated](https://github.com/testing-library/react-testing-library/pull/991) `renderHook` to the core `@testing-library/react` library. As stated in the [README](https://github.com/testing-library/react-hooks-testing-library#a-note-about-react-18-support):

> As part of the changes for React 18, it has been decided that the renderHook API provided by this library will instead be included as official additions to both react-testing-library ([PR](https://github.com/testing-library/react-testing-library/pull/991)) and react-native-testing-library ([PR](https://github.com/callstack/react-native-testing-library/pull/923)) with the intention being to provide a more cohesive and consistent implementation for our users.

With the introduction of [suspense](https://github.com/apollographql/apollo-client/pull/10323) to Apollo client, its become more obvious that we need to work with a version of `renderHook` that works with React 18 since it uses the new [concurrent rendering APIs](https://reactjs.org/blog/2022/03/08/react-18-upgrade-guide.html#updates-to-client-rendering-apis). To ensure Apollo client works with React 18 out of the box, it's imperative we run our React 18 tests in a way that we expect our users will use React 18.

There is one tricky issue though... with the migration of `renderHook` to the core library, we [can't run](https://github.com/testing-library/react-testing-library/issues/1121) our React 17 tests through it. The suggestion is to use `@testing-library/react^12` (which we do), but unfortunately the `renderHook` function is not exported from that version. We should be able to patch that behavior using [`patch-package`](https://www.npmjs.com/package/patch-package). In fact, the [suspense PR](https://github.com/apollographql/apollo-client/pull/10323/files#diff-50bd25fc977459fcebb6384986bb41ef9ab4e5300597523f533c0fe69b9140a7) does this for us. Once this change is in `main`, we can look to slowly update our tests over time to use hook from the core testing library.

We can use [this migration guide](https://github.com/testing-library/react-hooks-testing-library/blob/chore/migration-guide/MIGRATION_GUIDE.md) to migrate over when we are ready.

### PR DESCRIPTION
**Pull request by alessbell**
Closes #10333. Many thanks to @eps1lon for the foundation of this PR originating in https://github.com/eps1lon/apollo-client/pull/1 🙏

### To Do
- [x] uncomment and rewrite two tests in `hooks/__tests__/useQuery.test.tsx` which rely on [deprecated `result.all` API](https://github.com/testing-library/react-hooks-testing-library/blob/chore/migration-guide/MIGRATION_GUIDE.md#resultall)
- [x] fix 1 skipped test in `hooks/__tests__/useLazyQuery.test.tsx`
- [x] fix 1 failing test in React 18 (passing with 17) in `hooks/__tests__/useQuery.test.tsx`

### Checklist:

- [x] If this PR contains changes to the library itself (not necessary for e.g. docs updates), please include a changeset (see [CONTRIBUTING.md](../CONTRIBUTING.md#changesets))
- [x] If this PR is a new feature, please reference an issue where a consensus about the design was reached (not necessary for small changes)
- [x] Make sure all of the significant new logic is covered by tests

### CODE REVIEW

File: src/react/hooks/__tests__/useLazyQuery.test.tsx [hunk 1]
```diff
@@ -40,11 +42,13 @@ describe('useLazyQuery Hook', () => {
     const execute = result.current[0];
     setTimeout(() => execute());
 
-    await waitForNextUpdate();
-    expect(result.current[1].loading).toBe(true);
+    await waitFor(() => {
+      expect(result.current[1].loading).toBe(true);
+    }, { interval: 1 });
```
Comment (jerelmiller) [inline src/react/hooks/__tests__/useLazyQuery.test.tsx:47]: TIL about `interval`. Neat!

File: src/react/hooks/__tests__/useQuery.test.tsx [hunk 2]
```diff
@@ -3345,19 +3454,21 @@ describe('useQuery Hook', () => {
       );
 
       const onCompleted = jest.fn();
-      const { result, waitForNextUpdate } = renderHook(
-        () => useQuery(query, {
+      let updates = 0;
```
Comment (jerelmiller) [inline src/react/hooks/__tests__/useQuery.test.tsx:3457]: I'm not quite seeing where this `updates` variable is used. I see its written to in the `renderHook` call, but I don't see anything reading from it. Would you walk me through what this is used for?

File: src/react/hooks/__tests__/useQuery.test.tsx [hunk 3]
```diff
@@ -3686,18 +3813,20 @@ describe('useQuery Hook', () => {
...
-      await waitForNextUpdate();
-      // TODO: There is a missing update here because mutation and query update
-      // in the same microtask loop.
-      const previous = result.all[result.all.length - 2];
-      if (previous instanceof Error) {
-        throw previous;
-      }
+
+      expect(onError).toHaveBeenCalledTimes(0);
+      await waitFor(async () => {
```
Comment (jerelmiller) [inline src/react/hooks/__tests__/useQuery.test.tsx:3818]: [nit] You might consider using the `tick` helper here.

```ts
import { tick } from '../testing';

await tick();
```

**Review by jerelmiller — COMMENTED** [review]
Thank you SO SO much for doing this. Super great not to have to remember which `renderHook` we are using 😄.

Feel free to address/ignore the `[nit]` comments. There were a couple tests in here that might result in test flake, or false positives with the use of `waitFor`. I'm happy to approve once we determine if there is an issue there or not.

File: src/react/hooks/__tests__/useLazyQuery.test.tsx [hunk 4]
```diff
@@ -217,15 +227,17 @@ describe('useLazyQuery Hook', () => {
...
-    await waitForNextUpdate();
-    expect(result.current[1].loading).toBe(false);
+    await waitFor(() => {
+      expect(result.current[1].loading).toBe(false);
+    }, { interval: 1 });
     expect(result.current[1].data).toEqual({ hello: 'world 2' });
   });
 
-  it('should merge variables from original hook and execution function', async () => {
+  it.skip('should merge variables from original hook and execution function', async () => {
```
[the remark this answers is not in the document]
Comment (alessbell) [reply src/react/hooks/__tests__/useLazyQuery.test.tsx:240]: Yes, for now: it's one of the four tests still failing (I'll update the `To Do` list in the PR description as these are fixed 👍).

File: src/react/hooks/__tests__/useQuery.test.tsx (hunk 2 above)
Comment (alessbell) [reply src/react/hooks/__tests__/useQuery.test.tsx:3457]: Good catch! Removed 😄

File: src/react/hooks/__tests__/useLazyQuery.test.tsx [hunk 5]
```diff
@@ -1080,33 +1099,36 @@ describe('useLazyQuery Hook', () => {
...
+        expect(result.current.query.data).toBeUndefined();
+      }, { interval: 1 });
+
+      await waitFor(() => {
+        result.current.exec().then(result => {
+          expect(result.loading).toBe(false);
+          expect(result.called).toBe(true);
+          expect(result.data).toEqual({ counter: 1 });
+        });
+      }, { interval: 1 });
```
[the remark this answers is not in the document]
Comment (alessbell) [reply src/react/hooks/__tests__/useLazyQuery.test.tsx]: Oops, yes thanks for catching this! Turns out this was causing the React 17/18 discrepancy on the lines below: when the test is refactored to either return a value from `waitFor` or just await `result.current.exec()` before making assertions on the result, the conditional checks can be removed and a single test passes both React versions 🎉

**Comment by alessbell** [comment]
Thanks for the review, @jerelmiller! This is ready for a final pass—mostly minor fixes, with one call-out for the changes in https://github.com/apollographql/apollo-client/pull/10415/commits/6d1a204ee70a3423f08b166b8a8a986552af1c2b, the least straightforward test refactor in the PR.

**Review by jerelmiller — APPROVED** [review]
💯 💯 Nice work! Love when we can reduce dependencies. Thanks so much for doing this for us!

**Pull request merged**